### ISSUE TITLE
Cilium ingress with BGP control plane and TLS fails to serve any ingresses

### ISSUE THREAD

**Issue by samip5** [cilium/cilium#24146]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I deployed Cilium with BGP Control Plane and Ingress controller enabled and tried to setup an ingress with IPv6 and IPv4 enabled.

The ingress has IPv6 and IPv4 addresses attached to it, but IPv6 one doesn't respond at all (might be my own routing, not sure) and IPv4 one connects but fails in the TLS handshake.

### Cilium Version

cilium-cli: v0.12.1 compiled with go1.18.5 on linux/amd64
cilium image (default): v1.12.0
cilium image (stable): v1.13.0
cilium image (running): v1.13.0

### Kernel Version

`Linux w1 5.15.0-1024-raspi #26-Ubuntu SMP PREEMPT Wed Jan 18 15:29:53 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux`

### Kubernetes Version

```
{
  "clientVersion": {
    "major": "1",
    "minor": "26",
    "gitVersion": "v1.26.1",
    "gitCommit": "8f94681cd294aa8cfd3407b8191f6c70214973a4",
    "gitTreeState": "archive",
    "buildDate": "2023-02-06T14:43:47Z",
    "goVersion": "go1.19.4",
    "compiler": "gc",
    "platform": "linux/amd64"
  },
  "kustomizeVersion": "v4.5.7",
  "serverVersion": {
    "major": "1",
    "minor": "25",
    "gitVersion": "v1.25.6+k3s1",
    "gitCommit": "9176e03c5788e467420376d10a1da2b6de6ff31f",
    "gitTreeState": "clean",
    "buildDate": "2023-01-26T00:30:33Z",
    "goVersion": "go1.19.5",
    "compiler": "gc",
    "platform": "linux/arm64"
  }
}
```

### Sysdump

Too large; Uploaded here: https://cloud.skym.fi/index.php/s/dPK8a8QW2bqwoHr

### Relevant log output

```shell
level=info msg="LB-IPAM initializing" subsys=lbipam
level=info msg="Starting to synchronize CiliumNode custom resources" subsys=cilium-operator-generic
level=info msg="Starting to garbage collect stale CiliumNode custom resources" subsys=watchers
level=info msg="Starting to garbage collect stale CiliumEndpoint custom resources" subsys=cilium-operator-generic
level=info msg="Garbage collected status/nodes in Cilium Network Policies found=0, gc=0" subsys=cilium-operator-generic
level=info msg="Starting CRD identity garbage collector" interval=15m0s subsys=cilium-operator-generic
level=info msg="Garbage collected status/nodes in Cilium Clusterwide Network Policies found=0, gc=0" subsys=cilium-operator-generic
level=info msg="CiliumNodes caches synced with Kubernetes" subsys=cilium-operator-generic
level=info msg="Starting CNP derivative handler" subsys=cilium-operator-generic
level=info msg="Starting CCNP derivative handler" subsys=cilium-operator-generic
level=info msg="LB-IPAM done initializing" subsys=lbipam
level=info msg="Initialization complete" subsys=cilium-operator-generic
level=info msg="Start hook executed" duration=646.336068ms function="cmd.(*legacyOnLeader).onStart" subsys=hive
level=info msg="Notify ingress controller for service ingress" subsys=ingress-controller
level=error msg="Failed to process Ingress event, skipping: {ingress:0x40016b9e00}" subsys=ingress-controller
level=error msg="Failed to process Ingress event, skipping: {oldIngress:0x40016b9e00 newIngress:0x4000a18600}" subsys=ingress-controller
level=error msg="Failed to process Ingress event, skipping: {oldIngress:0x4000a18600 newIngress:0x4001288700}" subsys=ingress-controller
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by YutaroHayakawa** [comment]
At a glance, your BGP peer is never established. Could you check the peering state from your peer?

```
cat $(find . -type f -regex '\.\/logs-cilium-.*-cilium-agent-.*') | grep bgp-control-plane | grep ESTABLISHED
```

**Comment by samip5** [comment]
> At a glance, your BGP peer is never established. Could you check the peering state from your peer?
> 
> ```
> cat $(find . -type f -regex '\.\/logs-cilium-.*-cilium-agent-.*') | grep bgp-control-plane | grep ESTABLISHED
> ```

You're correct that the grep will not return anything, but:
```
2023-03-03T02:50:46.264783196+02:00 level=info msg="Peer Up" Key=10.0.110.1 State=BGP_FSM_OPENCONFIRM Topic=Peer asn=213021 component=gobgp.BgpServerInstance subsys=bgp-control-plane
2023-03-03T02:50:47.267666178+02:00 level=info msg="Peer Up" Key="2001:67c:1104:e110::1" State=BGP_FSM_OPENCONFIRM Topic=Peer asn=213021 component=gobgp.BgpServerInstance subsys=bgp-control-plane
```

And the other side confirms that.

```
k8s_dev_m1_v4 BGP        ---        up     2023-03-03 02:50:44  Established   
k8s_dev_m1_v6 BGP        ---        up     2023-03-03 02:50:42  Established   
k8s_dev_w1_v4 BGP        ---        up     2023-03-03 02:50:46  Established   
k8s_dev_w1_v6 BGP        ---        up     2023-03-03 02:50:47  Established
```

**Comment by youngnick** [comment]
The other side being Established is good, but if Cilium is still in OPENCONFIRM, then it's missed a keepalive message from the peer, so the peering is not established as far as Cilium is concerned. I haven't looked into that code at all yet, but I suspect that Cilium won't do a lot of stuff until the peer actually completes the BGP state transition and moves to Established.

Note that staying in OpenConfirm until a Keepalive message is received is part of the spec; a speaker is supposed to transition back to Idle if none is received. (Half-open peerings like this really suck to get working, sorry.)

**Comment by samip5** [comment]
> The other side being Established is good, but if Cilium is still in OPENCONFIRM, then it's missed a keepalive message from the peer, so the peering is not established as far as Cilium is concerned. I haven't looked into that code at all yet, but I suspect that Cilium won't do a lot of stuff until the peer actually completes the BGP state transition and moves to Established.
> 
> Note that staying in OpenConfirm until a Keepalive message is received is part of the spec; a speaker is supposed to transition back to Idle if none is received. (Half-open peerings like this really suck to get working, sorry.)

But if my other end is actually established (and receiving routes from Cilium) and Cilium just hasn't logged it being established which is what I was trying to get across?

**Comment by youngnick** [comment]
I'm sorry, I don't know any more than the command that @YutaroHayakawa has given you earlier, which seems to put your peers in OPENCONFIRM.

**Comment by YutaroHayakawa** [comment]
@samip5 I'm sorry for not getting back to you sooner. And currently, we only have a way to confirm the peering state from Cilium other than seeing the log. Based on what you said, your peer gets the route, so I suspect there's a problem in the forwarding layer, so I'd like to see the following next.

1. You said, `IPv6 one doesn't respond at all`, how did you test that? Can you do traceroute?
2. Based on the log output, you are using iBGP, means you may have multiple hops between Cilium node and the actual peer. How does your physical topology look like?
3. Can I see the nexthop of the route Cilium advertise?

**Comment by samip5** [comment]
@YutaroHayakawa

1. I tried an traceroute which works fine and curl even says connected.

curl example:
```
*   Trying [2001:67c:1104:e04e::ce55]:443...
* Connected to whoami.y2k.fi (2001:67c:1104:e04e::ce55) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
* Recv failure: Connection reset by peer
* OpenSSL SSL_connect: Connection reset by peer in connection to whoami.y2k.fi:443 
* Closing connection 0
curl: (35) Recv failure: Connection reset by peer
```

Outside trace:
```
Start: 2023-03-13T20:07:43+0200
HOST: lakka                                           Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- kapsi-gw1-vl1338.ip.kapsi.fi (2001:67c:1be8::2)  0.0%    10    0.4   2.3   0.2  20.0   6.2
  2.|-- 2a00:1d50:1:40::4:7605                           0.0%    10    0.8   0.8   0.5   2.1   0.5
  3.|-- 2a00:1d50:1::ce                                  0.0%    10    1.0   0.7   0.6   1.0   0.1
  4.|-- 2a00:1d50:1:24::6                                0.0%    10    2.8   3.4   2.7   8.6   1.8
  5.|-- core1.tku-fi.as213021.net (2a0b:a700:3::2)       0.0%    10    2.8   2.9   2.8   3.2   0.1
  6.|-- ???                                             100.0    10    0.0   0.0   0.0   0.0   0.0
  7.|-- ???                                             100.0    10    0.0   0.0   0.0   0.0   0.0
  8.|-- 2001:67c:1104:e04e::ce55                         0.0%    10   29.0  28.4  24.5  34.9   3.3
```

Inside trace (from BGP peer):
```
Start: 2023-03-13T20:09:57+0200
HOST: ER-12                    Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- 2001:67c:1104:d001::fb55  0.0%    10    0.5   0.5   0.4   0.5   0.0
  2.|-- 2001:67c:1104:e04e::ce55  0.0%    10    0.6   0.6   0.5   0.7   0.1
```

2. I tested it originally from a workstation, and it seems that currently the BGP is as established according to Cilium as well (Nothing else changed except I had rebooted the nodes and disabled/re-enabled BGP and Ingress)
3. Sure and these are for the Cilium ingress service:

```
Routing entry for 10.46.1.5/32
  Known via "kernel", distance 0, metric 32,  External Route Tag: 0, best
  * 10.0.110.19, via switch0.110
Routing entry for 2001:67c:1104:e04e::ce55/128
  Known via "kernel", distance 0, metric 32, best
  Last update 00:00:57 ago
  * via 2001:67c:1104:e110:e65f:1ff:fe1b:3293, switch0.110
```

**Comment by YutaroHayakawa** [comment]
Thanks! Now, the BGP is not an issue; it looks like the TLS handshake is failing for IPv6 only. So, for next, is it possible to provide us
1. The logs from the Cilium side
2. The TLS handshake parameters from the client side

@youngnick Let me ask you in parallel. Do you have any clue about the reason for the TLS handshake failure?

**Comment by samip5** [comment]
> 2. The TLS handshake parameters from the client side

I'm sorry but I don't quite understand what you mean by that?

**Comment by samip5** [comment]
> it looks like the TLS handshake is failing for IPv6 only

It's failing on both IPv4 and IPv6.

```
~ ❯❯❯ curl -v https://whoami.y2k.fi
*   Trying [2001:67c:1104:e04e::ce55]:443...
* Connected to whoami.y2k.fi (2001:67c:1104:e04e::ce55) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
* Recv failure: Connection reset by peer
* OpenSSL SSL_connect: Connection reset by peer in connection to whoami.y2k.fi:443 
* Closing connection 0
curl: (35) Recv failure: Connection reset by peer
~ ❯❯❯ curl -v -4 https://whoami.y2k.fi
*   Trying 10.46.1.5:443...
* Connected to whoami.y2k.fi (10.46.1.5) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
* Recv failure: Connection reset by peer
* OpenSSL SSL_connect: Connection reset by peer in connection to whoami.y2k.fi:443 
* Closing connection 0
curl: (35) Recv failure: Connection reset by peer
```

**Comment by youngnick** [comment]
Thanks @YutaroHayakawa for confirming it's not BGP, although I maintain that the OPENCONFIRM thing is weird. 😄 

Those "Failed to process Ingress event" messages do indicate that something is going wrong with Ingress processing, but I can't see anything obvious about the Ingress. @sayboras, do you have any thoughts?

**Comment by samip5** [comment]
Not stale.

**Comment by YutaroHayakawa** [comment]
@samip5 Hello, sorry for the late response. Is it still happening on the latest Cilium? Note that in the latest Cilium v1.13, we support checking the peering state with `cilium bgp peers` command. This gives you a better observation.

**Comment by rastislavs** [comment]
@samip5 In the past I have seen issues with similar behavior (TCP connection is established, but fails during the TLS handshake) that were related to MTU / fragmentation in the network (as soon as the server certificate is being sent, a large packet is generated and may be dropped in case of MTU issues). It would be very helpful if you could collect tcpdump traces of a failed attempt (or better both - IPv4 and IPv6 - you can use curl's `--ipv4` & `--ipv6` params to enforce one or the other) at both client side and cilium node side to check that.

**Comment by samip5** [comment]
> @samip5 In the past I have seen issues with similar behavior (TCP connection is established, but fails during the TLS handshake) that were related to MTU / fragmentation in the network (as soon as the server certificate is being sent, a large packet is generated and may be dropped in case of MTU issues). It would be very helpful if you could collect tcpdump traces of a failed attempt (or better both - IPv4 and IPv6) at both client side and the cilium node side to check that.

I believe the issue here is that it simply doesn't know how to handle it when it doesn't have a valid certificate yet but you ask it to use https.

**Comment by rastislavs** [comment]
> I believe the issue here is that it simply doesn't know how to handle it when it doesn't have a valid certificate yet but you ask it to use https.

Ah right, I assumed that the ingress was properly configured including TLS certificate. I just tried an ingress without proper TLS config and it behaves exactly like that:

```
curl --insecure -v https://bookinfo.cilium.rocks 
*   Trying 192.168.10.229:443...
* Connected to bookinfo.cilium.rocks (192.168.10.229) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* OpenSSL SSL_connect: Connection reset by peer in connection to bookinfo.cilium.rocks:443 
* Closing connection 0
curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to bookinfo.cilium.rocks:443 
```

As soon as I properly configured TLS in the ingress resource, this started working. 

@samip5 Could you please check if this works if you set up TLS e.g. [based on this example](https://docs.cilium.io/en/stable/network/servicemesh/tls-termination/#create-tls-certificate-and-private-key)?

BTW, a good way to debug the connectivity issues on Envoy level is to turn on debugging [like this](https://docs.cilium.io/en/stable/operations/troubleshooting/#connectivity-troubleshooting):
```
$ kubectl get -n kube-system cm cilium-config -o json | grep "debug"
        "debug": "true",
        "debug-verbose": "flow",
```
With this, you could spot the reason why the connection is closed on Envoy, e.g. this was the case for me with missing TLS config:
```
level=debug msg="[Cilium SocketOption(): source_identity: 8, ingress: false, port: 443, pod_ip: 10.244.2.190, source_addresses: /10.244.2.190:0/[fd00:10:244:2::cb2e]:0, mark: 80b00 (magic mark: b00, cluster: 0, ID: 8)" subsys=envoy-filter threadID=793
level=debug msg="[closing connection from 172.18.0.5:41980: no matching filter chain found" subsys=envoy-conn_handler threadID=793
```

**Comment by samip5** [comment]
Coming from ingress-nginx, the bahavior regarding not accepting a connection at all without a ready cert, seems odd.

ingress-nginx let's specify the default cert and use that.

**Comment by rastislavs** [comment]
> Coming from ingress-nginx, the bahavior regarding not accepting a connection at all without a ready cert, seems odd.
> 
> ingress-nginx let's you specify the default cert and use that instead in cases like this.

Yeah, I agree that from usability perspective it could be better. The point is that if this is the only issue, we can definitely rule out the BGP control plane, and this potentially could be addressed as an enhancement / CFP for the Service Mesh. Could you please confirm that @samip5 ?

**Comment by samip5** [comment]
Yes, BGP control plane works fine, despite when it said incorrectly that it wasn't established in the start of the issue.

**Comment by sayboras** [comment]
:wave: 

Recently, we have added the support for default cert, much appreciated if you can test it out and confirm if there is still any issue. 

https://github.com/cilium/cilium/pull/26065

**Comment by samip5** [comment]
I have gotten around to testing this and can verify it works with the default cert so I'm closing this.

**Comment by aristeidist** [comment]
Have the same issue but not using BGP.

Cilium install options are:
helm install cilium cilium/cilium --version 1.15.5 \
    --namespace kube-system \
    --set l2announcements.enabled=true \
    --set kubeProxyReplacement=true \
    --set k8sServiceHost=${API_SERVER_IP} \
    --set k8sServicePort=${API_SERVER_PORT} \
    --set ingressController.enabled=true \
    --set ingressController.loadbalancerMode=dedicated \
    --set externalIPs.enabled=true \
    --set hubble.relay.enabled=true

Cilium:
cilium-cli: v0.16.9 compiled with go1.22.3 on linux/amd64
cilium image (default): v1.15.5
cilium image (stable): v1.15.5
cilium image (running): 1.15.5

Running on Bare metal:
2 VM's on Proxmox x86 (Including Control plane)
1 VM on Raspberry (I have **apt install linux-modules-extra-*** )
All network connectivity tests are ok by cilium
Using Cilium L2 IPAM 

I get:
```
* ALPN: curl offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* Recv failure: Connection reset by peer
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* OpenSSL SSL_connect: Connection reset by peer in connection to [IP]:443
* Closing connection
curl: (35) Recv failure: Connection reset by peer
```

If I install nginx  no issues with same certificates this is my ingress description:
```
Name:             cil-hubui-ingress-tls
Labels:           <none>
Namespace:        kube-system
Address:          [REALIPPROVIDEDBY]
Ingress Class:    cilium
Default backend:  <default>
TLS:
  [secretename] terminates [mydomainname]
Rules:
  Host                   Path  Backends
  ----                   ----  --------
  [mydomainname]
                         /   hubble-ui:80 (10.0.3.39:8081)
Annotations:             <none>
Events:                  <none>
```
This is my hubble-ui example:
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: cil-hubui-ingress-tls
  namespace: kube-system
  annotations:
    #ingress.cilium.io/service-type: "NodePort"
    #ingress.cilium.io/tls-passthrough: "enabled"
spec:
  ingressClassName: cilium
  rules:
  - host: [HOSTNAME]
    http:
      paths:
      - backend:
          service:
            name: hubble-ui
            port:
              number: 80
        path: /dash
        pathType: Prefix
  tls:
  - hosts:
    - [HOSTNAME]
    secretName: [secretename]

```
I see no issues or IP allocation, or routing at least based on the fact that nginx ingress works and cilium ingress does not and gives an error with same certificate configuration.

Any help would be appreciated here.

**Comment by thecloudgarage** [comment]
I am facing the exact same issue... NGINX INGRESS CONTROLLER works perfect with Cilium BGP control plane.. My https app with an Ingress resource with ingress class nginx works perfect.. the nginx ingress controller receives the ip from LB IPAM and things are perfect .  The same configuration with Cilium ingress don't work.. no changes... Cilium ingress is enabled and receives IP from LB IPAM... INGRESS resource class changes from nginx to Cilium... kubectl get ingress shows the loadbalancer address... but when I try to access ... it results in an exact same OpenSsl error given above

**Comment by aristeidist** [comment]
I kwow...I gave up and went to nginx ingress to start with new issues with wp-admin

**Comment by thecloudgarage** [comment]
That's very SAD!!!!! Is there any response from Cilium team?????

**Comment by thecloudgarage** [comment]
FINALLY!!! After days of scavenging and no results., decided to turn off cilium config set enable-ipv4-masquerade false and voila.. it works... TLS certs with Cilium Ingress and BGP control plane...

**Comment by aristeidist** [comment]
Wow, thanks for the find, will use it for my cilium ingress and report back
